pkg/instance/pool.go

```diff
@@ -0,0 +1,277 @@
+
+package instance
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/vm"
+)
+
+type Instance interface {
+	Index() int
+	Close()
+}
+
+type UpdateInfo func(cb func(info *Info))
+type Runner[T Instance] func(ctx context.Context, inst T, updInfo UpdateInfo) error
+type CreateInstance[T Instance] func(int) (T, error)
+
+// Pool[T] provides the functionality of a generic pool of instances.
+// The instance is assumed to boot, be controlled by one Runner and then be re-created.
+// The pool is assumed to have one default Runner (e.g. to be used for fuzzing), while a
+// dynamically controlled sub-pool might be reserved for the arbitrary Runners.
+type Pool[T Instance] struct {
+	BootErrors chan error
+
+	count      int
+	creator    CreateInstance[T]
+	defaultJob Runner[T]
+	jobs       chan Runner[T]
+
+	// The mutex serializes ReserveForRun() calls.
+	mu        sync.Mutex
+	instances []poolInstance[T]
+}
+
+func NewVMPool(pool *vm.Pool, def Runner[*vm.Instance]) *Pool[*vm.Instance] {
+	return NewPool(pool.Count(),
+		func(idx int) (*vm.Instance, error) {
+			return pool.Create(idx)
+		},
+		def)
+}
+
+func NewPool[T Instance](count int, creator CreateInstance[T], def Runner[T]) *Pool[T] {
+	instances := make([]poolInstance[T], count)
+	for i := 0; i < count; i++ {
+		instances[i].idx = i
+		instances[i].free(def)
+	}
+	return &Pool[T]{
+		BootErrors: make(chan error, 16),
+		count:      count,
+		creator:    creator,
+		defaultJob: def,
+		instances:  instances,
+		jobs:       make(chan Runner[T]),
+	}
+}
+
+func (p *Pool[T]) Loop(ctx context.Context) {
+	var wg sync.WaitGroup
+	wg.Add(p.count)
+	for i := 0; i < p.count; i++ {
+		i := i
+		go func() {
+			for ctx.Err() == nil {
+				p.runInstance(ctx, i)
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+}
+
+func (p *Pool[T]) runInstance(ctx context.Context, idx int) {
+	obj := &p.instances[idx]
+	ctx, cancel := context.WithCancel(ctx)
+
+	log.Logf(2, "pool: booting instance %d", idx)
+
+	p.mu.Lock()
+	// Avoid races with ReserveForRun().
+	obj.reset(cancel)
+	p.mu.Unlock()
+
+	obj.status(StateBooting)
+	defer obj.status(StateOffline)
+
+	inst, err := p.creator(idx)
+	if err != nil {
+		p.BootErrors <- err
+		return
+	}
+	defer inst.Close()
+
+	obj.status(StateWaiting)
+	// The job and jobChan fields are subject to concurrent updates.
+	obj.mu.Lock()
+	job, jobChan := obj.job, obj.jobChan
+	obj.mu.Unlock()
+
+	if job == nil {
+		select {
+		case newJob := <-jobChan:
+			job = newJob
+		case newJob := <-obj.switchToJob:
+			job = newJob
+		case <-ctx.Done():
+			return
+		}
+	}
+
+	obj.status(StateRunning)
+	job(ctx, inst, obj.updateInfo)
+}
+
+func (p *Pool[T]) ReserveForRun(VMs int) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if VMs > p.count {
+		panic("trying to reserve more VMs than present")
+	}
+
+	var free, reserved []*poolInstance[T]
+	for i := 0; i < p.count; i++ {
+		inst := &p.instances[i]
```

Comment:
`p.count` duplicates `len(p.instances)`, unclear which is the source of truth.
Perhaps we should make instances a slice of _poitners_ and then use:
```
	for _, inst := range p.instances {
```
or
```
	for i := range p.instances {
```